Remove type-distinction between system-communicators and user-communicators

Cydhra opened the pull request:
As outlined in #148, it is very unintuitive to split the Communicator interface in `SystemCommunicator` and `UserCommunicator`.
I drafted a PR to unify both into one common `enum SimpleCommunicator`:

- I migrated topology-related code from UserCommunicator into the common enum, so both communicator variants can now call those functions. I couldn't confirm it outright, but the spec does not prohibit that explicitely, so I think this models the MPI interface more closely.
- I refactored CartesianCommunicator to accept both World- and UserCommunicator, which is legal by spec, I don't know why this wasn't possible before.
- since WorldCommunicator is no longer Copy, code that moved a reference to SystemCommunicator must now explicitly refer to IntraCommunicator::WorldCommunicator (see immediate_multiple_test.rs)
- I added SelfCommunicator next to WorldCommunicator, because the MPI spec defines it and without it downstream crates would need unsafe code to create it, and call `free` on it when their `UserCommunicator` drops

Things missing from this draft:

There is another weird split in the code-base: InterCommunicators use `SystemCommunicatorHandle` and `UserCommunicatorHandle`, even though those handle structs do the exact same as their communicator-counterparts (at least as far as I can tell).
Since I have never worked with InterCommunicators before though, and there are neither integration-tests nor examples for them, I didn't migrate them to use `SimpleCommunicator` yet, as I wanted to get some feedback on their purpose.

All examples and tests are green, but I am not very familiar with the codebase yet, so somebody should scrutinize this.

jedbrown commented:
Thanks! The `spawn` examples use inter-communicators. A subtle point is that the inter-comm returned by `MPI_Comm_get_parent` should be thought of as a "system" inter-comm since freeing it would leave other references dangling. This is in contrast to the inter-comm returned by `MPI_Comm_spawn`, which has normal "user" semantics. This might help explain why the separate handles exist. I'm not attached to this design, BTW.

```rs
/// A system-defined inter-communicator
pub type SystemInterCommunicator = InterCommunicator<SystemCommunicatorHandle>;

/// A user-defined inter-communicator
pub type UserInterCommunicator = InterCommunicator<UserCommunicatorHandle>;
```

Looks like we have some CI failures.

Cydhra commented:
> A subtle point is that the inter-comm returned by `MPI_Comm_get_parent` should be thought of as a "system" inter-comm since freeing it would leave other references dangling. This is in contrast to the inter-comm returned by `MPI_Comm_spawn`, which has normal "user" semantics. This might help explain why the separate handles exist. I'm not attached to this design, BTW.

The reason why the separate handles exist is the same why `SystemCommunicator` and `UserCommunicator` were split in the first place, I suppose. My question was why use a separate -`Handle` instead of `SystemCommunicator` and `UserCommunicator`.

I don't quite understand if Drop-semantics are important here: Intuitively I wouldn't even unify the two handles into a common `SimpleCommunicatorHandle`, but instead use `SimpleCommunicator` directly. But I don't know if there was a reason to differentiate between `*CommunicatorHandle` and `*Communicator`.

> Looks like we have some CI failures.

I was bitten by `#[cfg(not(msmpi))]`, since I tested everything on Windows. Fixing it on Fedora right now.

jedbrown commented on src/topology/mod.rs:
```diff
@@ -246,74 +148,123 @@ impl UserCommunicator {
-        cart_comm.0
-    }
-}
+/// A handle to a communicator used in InterCommunicators. This type is functionally similar to `SimpleCommunicator`
+/// save for its Drop-semantics.
+pub enum SimpleCommunicatorHandle {
```
I think we have a weird middle ground here where there is a type-level distinction between system and user inter-communicators, but not for intra-communicators. `Self` and `World` are not inter-communicators.

`MPI_Com_get_parent` returns a "system" inter-comm, with possibly surprising semantics.

![image](https://user-images.githubusercontent.com/3303/225696679-b02ee242-273c-44b5-ae3a-41c72943f428.png)

I think we could make `InterCommunicator` an `enum` and remove `*Handle`s entirely. Or we could have more privacy by making a single `enum CommunicatorHandle` and having a non-public (perhaps `pub(crate) CommunicatorHandle`) field in `SimpleCommunicator`, `IntraCommunicator`, and the topology communicators. So all handling of a raw `MPI_Comm` would be inside `CommunicatorHandle` and yet we wouldn't be revealing those publicly. This would prevent users from pattern matching on whether a given communicator is `World` (for example), but that's a code smell anyway, and it would prevent future internal changes from breaking user code. What do you think?

Cydhra replied:
The reason `World` is still in this enum is that I assume `MPI_COMM_GET_PARENT` can return WORLD as a parent. From the spec, it sounds like this cannot happen, but the codebase explicitely defined an InterComm with `SystemCommunicatorHandle`, so I assumed that it is necessary. 

If it is, then making the `Handle` an `enum` with `pub(crate)` fields in the relevant structs is probably the most elegant solution. We would remove the `Drop` implementation from the `Handle` and instead add it to the public wrapper types `SimpleCommunicator`, `InterCommunicator` and the topology communicators.

If all of those communicators implement `Communicator` and this trait then contains all methods like it was initially suggested in #148, the split in the type system is gone completely. My solution so far only removes it between `System*` and `User*`, not between intra- and intercommunicators.

jedbrown replied:
`MPI_Comm_get_parent` cannot return `WORLD`. We had `InterCommunicator<SystemCommunicatorHandle>` so that `world.parent()` can be called multiple times without returning None or leaving dangling references.

I was thinking that `Handle` would implement dropping semantics so that code doesn't clutter the higher levels (public types). I think `SimpleCommunicator`, `InterCommunicator`, and the topological communicators would be publicly visible types because they have some arguments that don't apply to general communicators. I haven't thought about whether we need an `IntraCommunicator` trait (implemented by `SimpleCommunicator` and the topological comms) so it can be used as a type constraint.

Cydhra replied:
If the `Handle` is used in both Inter- and Intracomms, Drop cannot be implemented on the handle, because it must either call `MPI_COMM_DISCONNECT` or `MPI_COMM_FREE`.

jedbrown replied:
If `Handle` is an enum, its `drop` implementation can match on whether to do nothing (World, Self, and parent), to `MPI_Comm_disconnect` as with a child inter-comm, or to `MPI_Comm_free` (user intra-comm). The constructors would need to create the correct handle before putting it in a public type, but then nothing else would need to implement `drop`.

Cydhra replied:
Sorry for the delay, I am amidst exam season. I will  try to implement the second option, turning the Handle into an enum and hiding it from the public API. As you were saying, this reduces code smell by smoothing the distinction between World- and User-comms, reduces the overall API complexity, and hides away the raw handles properly. This doesn't address the initial concern of a unified `Communicator` trait yet, but as suggested in the issue we could then probably move everything into `trait Communicator` and use `dyn Communicator` for downstream crates.

The only problem I currently see are immediate requests, because they consume the communicator, which may introduce new complexity once the raw communicator is hidden away. I haven't looked into that yet, I assume this has to do with RAII semantics. I will ignore it for now, but this issue needs to be looked into and fixed before merging.

Cydhra commented on examples/immediate_multiple_test.rs:
```diff
@@ -67,7 +76,15 @@ fn main() {
+            prev_proc,
+            &x,
+            &mut recv,
+        );
```
Oh yeah that is probably the better solution for this particular case. I wasn't thinking this far because when I saw that the previous code broke, I worried about the *double free* problem.

Cydhra commented:
I have implemented an enum `CommunicatorHandle` now, deleted the previous handles and their trait, simplified the API to one struct `SimpleCommunicator`, one `InterCommunicator` and the topological communicators.

The latter still use simple communicators in their constructor, I don't know if that is how it is supposed to be, because it seems weird to have this conversion going on. I need some feedback on that.

The Handle type is still public, I don't know if we should hide it completely. At the moment hiding it is weird, because there is the public trait `AsRaw` that exposes the raw handles from ffi anyway. If we want to hide the handle enum, we should also hide `AsRaw`.

jedbrown commented:
Cool, I'll check it out as time allows.

I think of `AsRaw` as public because a user may want to call a C function (say PETSc) that takes an `MPI_Comm` (relates to #32). My current thinking is that `AsRaw` can stay public while hiding the handle. One should be able to create any of the concrete implementations from a raw `MPI_Comm`, but I don't think we need to see the `CommunicatorHandle` for that either. Do you have a use case?

Cydhra commented:
> Cool, I'll check it out as time allows.
> 
> I think of `AsRaw` as public because a user may want to call a C function (say PETSc) that takes an `MPI_Comm` (relates to #32). My current thinking is that `AsRaw` can stay public while hiding the handle. One should be able to create any of the concrete implementations from a raw `MPI_Comm`, but I don't think we need to see the `CommunicatorHandle` for that either. Do you have a use case?

Nope. At the moment `InterCommunicator` can be created from `CommunicatorHandle` (which is a bit weird) so this needs to be changed to hide the handle. I wanted to ask whether to hide the handle before changing it, because it was like this before. For some reason the check whether a raw handle is of an inter-comm was implemented in the InterComm constructor instead of the Handle-constructor. I have since changed that.

jedbrown commented on src/topology/cartesian.rs:
```diff
@@ -51,8 +52,8 @@ impl CartesianCommunicator {
     /// - `raw` must not be used after calling `from_raw`.
-    pub unsafe fn from_raw(raw: MPI_Comm) -> Option<CartesianCommunicator> {
-        UserCommunicator::from_raw(raw).and_then(|comm| match comm.into_topology() {
+    pub unsafe fn from_raw_checked(raw: MPI_Comm) -> Option<CartesianCommunicator> {
```
Why did `from_raw` change from returning an `Option` to an unchecked version? Can we avoid that without contortions? In general, I'd rather discourage using unchecked versions.

Cydhra replied:
Because I implemented `FromRaw` for everything and that already used the name `from_raw` for the unchecked version, so I renamed the checked version of the function.

Cydhra replied:
If you are worried about breaking changes: Since we remove types like `SystemCommunicator` we are already breaking. And this won't silently change semantics since the return-type changed.

Would a mention of the checked version in the docs suffice, or should we hide the unchecked conversions from public API entirely, or should I revert those changes and declare it out-of-scope for this PR?

jedbrown replied:
I'm not that worried about the breaking change (so long as it's easy for users to update), but rather consistency and encouraging safer use. When in doubt, people usually choose the shorter version of the function. Looking at core, it seems the patterns could be `from_raw, try_from_raw` or `from_raw_unchecked, from_raw`. I guess either is okay with me. I didn't see much prior art for the `from_raw_checked` naming, but maybe you have.

jedbrown reviewed (commented):
Thanks for your work here and sorry for my slow review. Would it be a lot to ask for an example showing something that can be done in the new model that couldn't be done before?

I suspect the new version can compile a little faster/smaller due to fewer variants to monomorphize, though even that is hard to measure. I find it a bit more intuitive, but it seems to be almost exactly the same amount of code. Unfortunately, we can still hardly use `&dyn Communicator` because most methods require `Self: Sized`. I'm not sure how to gracefully get us out of that bind.

Cydhra commented on src/topology/mod.rs:
```diff
@@ -54,159 +54,213 @@ pub trait AsCommunicator {
-}
+    /// Create a `CommunicatorHandle::InterCommunicator` from a raw handle. Returns `None` if the
+    /// handle is `MPI_COMM_NULL`.
+    pub fn inter_from_raw(raw: MPI_Comm) -> Option<CommunicatorHandle> {
```
Probably? The use of `unsafe` in the API does not follow a pattern I could recognize. I didn't use unsafe when the function performed checks and returned an `Option` because that is largely how it was implemented before my changes.

Cydhra commented on src/topology/mod.rs:
```diff
@@ -54,159 +54,213 @@ pub trait AsCommunicator {
-        UserCommunicator(raw)
+        let handle = CommunicatorHandle::simple_from_raw_unchecked(raw);
+        SimpleCommunicator(handle)
     }
```
I agree. The problem is, I tried to minimize the havoc I caused: I initially started this PR, because I wanted to unify the communicators, which would side-step the `&dyn Communicator` problem. I didn't realize that inter-communicators behaved differently, so this endeavor failed.

I continued to at least clean up the code-base by unifying the raw handle like you suggested, but I didn't want to blindly remove functionality that was there before. I think we shouldn't expose the unchecked constructors of any structure to public API, but I tried not to disrupt the API too much. I will gladly go over it again though, to check which unchecked constructors can be removed, and what checks can be added to the remaining constructors, if that is desired.

To this end, there is another question though: `FromRaw` is a trait, for which I retrofitted implementations into the communicator and handle structs, because you wanted downstream crates to be able to call them with raw handles (and it seemed stupid to not use the already existing trait, even if it wasnt implemented before). I used unsafe and unchecked implementations for those, because the trait does not return an `Option` and was already marked `unsafe`. Should this trait and its implementations remain, and/or should there be a safe alternative trait?

Cydhra replied:
Okay so, I implemented the `FromRaw` trait for nearly everything (because you wanted a way for users to create the types from raw handles) and also tried to provide `try_from_raw` versions everywhere.

Should I get rid of unchecked versions (implemented via the `FromRaw` trait) where we don't need them, or is it fine the way it is (it mostly was like this before).

Cydhra commented:
About the example what can be done in the new model:
I don't think it changes that much to be honest. I initially wanted to change it, because I was using `&SystemCommunicator` in my crate, which was overly restrictive. So I tried to unify it with `UserCommunicator` (because as #148 points out, `dyn Communicator` doesnt work), but then realized that I cannot easily unify the struct with `InterCommunicator`.

I then changed my approach to your suggestion of at least encapsulating the handle correctly, which further disconnected this PR from its initial attempt to make it easier to support different communicators for higher-level routines: Now the distinction between Intra-, Inter- and topological communicators still exists publicly, only the handle is nicely modelled. I am still hoping this will simplify the solution for #148, but from the outside, this PR doesn't improve the API that much, sadly.

About all the safety-related stuff: I tried to not create more mess than was already present, but some functions were and are not marked unsafe that really should be (then again, at least some collective communication functions are unsound but not marked unsafe either).

Cydhra commented:
Is this still desired?

jedbrown commented:
Sorry, I got swamped with end-of-semester stuff. I'll be able to put more attention toward this next week.

jedbrown reviewed (commented):
Thanks. Sorry about losing track of this. Pipeline is running now.
